src/api/auth/auth.cpp

```diff
@@ -103,6 +103,13 @@ static void TryAuthenticate(
 	const string signature,
 	APIResponseHandler api_handler);
 
+static void callAPiHandler(
+	mender::http::Client &client, APIResponseHandler api_handler, ExpectedAuthData exp) {
+	// Explicitly cancel the HTTP client to ensure that the socket gets closed
+	api_handler(exp);
+	client.Cancel();
```

Comment:
A couple of things:

* I would also comment that we are not interested in the body.
* It might better to cancel before calling the handler, if possible, because the handler may start new requests.
* Keep in mind that cancelling may trigger cancellation messages elsewhere in the code, so make sure you try this in practice a bit.

Reply:
> A couple of things:
> 
>     * I would also comment that we are not interested in the body.

We do read the body. We were miss-lead yesterday when we looked at the code together but the whole body is read (and the json response from the server is logged) before calling the user.

>     * It might better to cancel before calling the handler, if possible, because the handler may start new requests.
> 

I tried this first, but the tests would segfault.

The segfault and the fact that the reader(s) are not properly destroyed made me suspect that the issue is the resistivity of this code, where a failed auth triggers a new fetch (with the potential next server url) passing around the handlers. In other words I would expect that we don't need to do `client.Cancel()` here but I couldn't find a different way to fix this problem.

Reply:
> We do read the body. We were miss-lead yesterday when we looked at the code together but the whole body is read (and the json response from the server is logged) before calling the user.

We do? But then this is not the appropriate fix. It means there is a bug somewhere which prevents closing the socket on a completed request. We should find where that is so it can be fixed centrally.

Reply:
Yeah, to me having to call `client.Cancel()` here explicitly looks like a bug.

Reply:
The bug was indeed somewhere else. Rework coming up :eyes:

Reply:
Thank you both for being skeptical about this "fix". A different approach is coming up in the next force-push :eyes: